ginivision/src/main/java/net/gini/android/vision/internal/fileimport/FileChooserActivity.java

```diff
@@ -227,15 +241,102 @@ private void populateFileProviders() {
                 new ProvidersAppItemSelectedListener() {
                     @Override
                     public void onItemSelected(@NonNull final ProvidersAppItem item) {
-                        final Intent intent = item.getIntent();
-                        intent.setClassName(
-                                item.getResolveInfo().activityInfo.packageName,
-                                item.getResolveInfo().activityInfo.name);
-                        startActivityForResult(intent, REQ_CODE_CHOOSE_FILE);
+                        LOG.info("Requesting read storage permission");
+                        requestStoragePermission(new PermissionRequestListener() {
+                            @Override
+                            public void permissionGranted() {
+                                LOG.info("Read storage permission granted");
+                                launchApp(item);
+                            }
+
+                            @Override
+                            public void permissionDenied() {
+                                LOG.info("Read storage permission denied");
+                                showStoragePermissionDeniedDialog();
+                            }
+
+                            @Override
+                            public void shouldShowRequestPermissionRationale(
+                                    @NonNull final RationaleResponse response) {
+                                LOG.info("Show read storage permission rationale");
+                                showStoragePermissionRationale(response);
+                            }
+                        });
+
+
                     }
                 }));
     }
 
+    private void launchApp(final @NonNull ProvidersAppItem item) {
+        final Intent intent = item.getIntent();
+        intent.setClassName(
+                item.getResolveInfo().activityInfo.packageName,
+                item.getResolveInfo().activityInfo.name);
+        startActivityForResult(intent, REQ_CODE_CHOOSE_FILE);
+    }
+
+    private void showStoragePermissionDeniedDialog() {
+        final AlertDialog alertDialog = new AlertDialog.Builder(this)
+                .setMessage(R.string.gv_storage_permission_denied)
+                .setPositiveButton(R.string.gv_storage_permission_denied_positive_button,
+                        new DialogInterface.OnClickListener() {
+                            @Override
+                            public void onClick(final DialogInterface dialog, final int which) {
+                                LOG.info("Open app details in Settings app");
+                                showAppDetailsSettingsScreen();
+                            }
+                        })
+                .setNegativeButton(R.string.gv_storage_permission_rationale_negative_button, null)
+                .create();
+        alertDialog.setCanceledOnTouchOutside(false);
+        alertDialog.show();
```

Comment:
`AlertDialog` doesn't survive configuration changes(e.g. orientation change, don't keep activities), might make sense to use a `DialogFragment`.